### PR TITLE
Dev joris#472

### PR DESCRIPTION
**Pull request by JorisBassle**
image

Also PR in the backend

### CODE REVIEW

File: src/components/GroupManagement.vue [hunk 1]
```diff
@@ -252,15 +298,34 @@
         });
     });
 
+    const changeUserRole = (contact: GroupContact | Contact) => {
+        if ('roles' in contact) {
```
Comment (JonasWijne) [inline src/components/GroupManagement.vue:302]: should be changed

File: src/components/GroupManagement.vue [hunk 2]
```diff
@@ -252,15 +298,34 @@
...
+            selectedUser.value = contact;
+        }
+    };
+
+    const doChangeUserRole = () => {
+        const { updateContactsInGroup } = usechatsActions();
+
+        if (selectedUser.value.roles.includes(Roles.MODERATOR)) {
+            selectedUser.value.roles.pop();
+        } else selectedUser.value.roles.push(Roles.MODERATOR);
```
Comment (JonasWijne) [inline src/components/GroupManagement.vue:313]: no else in my codebase

File: src/components/GroupManagement.vue [hunk 3]
```diff
@@ -274,12 +339,19 @@
...
+
     const isAdmin = computed(() => {
-        const { user } = useAuthState();
         //@ts-ignore
         return props.chat.adminId == user.id;
     });
 
+    const isModerator = computed(() => {
+        const contact = props.chat.contacts.find(c => c.id === user.id);
+        if ('roles' in contact) return contact.roles.includes(Roles.MODERATOR);
```
Comment (JonasWijne) [inline src/components/GroupManagement.vue:351]: should be changed

File: src/store/chatStore.ts [hunk 4]
```diff
@@ -564,6 +566,10 @@ const updateContactsInGroup = async (
     let msg = `${contact.id} has been removed from the group`;
     if (type === SystemMessageTypes.ADD_USER) msg = `${contact.id} has been added to the group`;
     if (type === SystemMessageTypes.USER_LEFT_GROUP) msg = `${contact.id} has left the group`;
+    if (type === SystemMessageTypes.CHANGE_USER_ROLE) {
+        const roleChange = contact.roles.includes(Roles.MODERATOR) ? 'promoted' : 'demoted';
+        msg = `${contact.id} has been ${roleChange}`;
+    }
```
Comment (JonasWijne) [inline src/store/chatStore.ts:572]: should be a switch and maybe move to a different function

File: src/store/socketStore.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
-import { Chat, Id, Message } from '@/types';
+import { Chat, GroupContact, Id, Message, Roles } from '@/types';
 import { reactive } from '@vue/reactivity';
-import { toRefs, inject } from 'vue';
+import { inject, toRefs } from 'vue';
 import { handleRead, removeChat, usechatsActions } from './chatStore';
```
Comment (JonasWijne) [inline src/store/socketStore.ts:4]: are these used ?

**Review by JonasWijne — CHANGES_REQUESTED** [review]

File: src/store/socketStore.ts (hunk 5 above)
Comment (JorisBassle) [reply src/store/socketStore.ts:4]: yes
image
image

**Review by JonasWijne — APPROVED** [review]

**Pull request merged**